### ISSUE TITLE
[Bug]: The order of type definitions influences mapping subtype relation

### ISSUE THREAD

**Issue by lochana-chathura** [ballerina-platform/ballerina-lang#42787]
### Description

Consider the following cases, 

- Case 1
```bal
type M map<any>;
type N map<any>;

// @type M1 = M2
type M1 M|record {};

type M2 N|record {};
```

- Case 2
```bal
// @type M1 = M2
type M1 M|record {};

type M2 N|record {};

type M map<any>;
type N map<any>;
```

- Case 3
```bal
type M map<any>;

// @type M1 = M2
type M1 M|record {};

type M2 map<any>|record {};
```

- Case 4
```bal
// @type M1 = M2
type M1 map<any>|record {};

type M2 map<any>|record {};
```

Although we have defined the same types in all 4 cases. The subtyping relation is different. 
1. Results in M1 <> M2. So, the test **fails**.
2. Results in M1 = M2. So, the test **passes**.
3. Results in M1 <: M2. So, the test **fails**.
4. Results in M1 = M2. So, the test **passes**.

In cases 2 and 4, the `record {}` type gets defined first, therefore having a lower index compared to `map<any>` type. This influences the` isEmpty()` check on top of the `diff()`.

**Comment by lochana-chathura** [comment]
The BDD for `diff(M1, M2)` is as follows.

For the failing cases,
<img width="895" alt="Screenshot 2024-05-20 at 18 27 48" src="https://github.com/ballerina-platform/ballerina-lang/assets/39232462/b222454b-6020-41be-8741-13dbfbeb66c1">

For the passing cases, 
The tree is same with `map<any>` and `map<anydata>` interchanged.

**Comment by lochana-chathura** [comment]
Overriding `MappingAtomicType` and `ComplexSemType`s' `equals()` and `hashCode()` solved the issue for the original sample. However the following still fails as deep equality is not considered in `subtypeDataList`.

```
type M map<(1|2|3)>;
type N map<(1|2|3)>;

// @type M1 = M2
type M1 M|map<(1|2)>;

type M2 N|map<(1|2)>;
```

**Comment by lochana-chathura** [comment]
I was able to reproduce the same scenario(for the original sample) in nBallerina by modifying `mappingAtom()` as follows.

 ```bal
    isolated function mappingAtom(MappingAtomicType atomicType) returns TypeAtom {
        lock {
            TypeAtom? ta = self.atomTable[atomicType];
            if ta != () {
                TypeAtom result = { index: self.atomTable.length(), atomicType };
                return result;
            }
            else {
                TypeAtom result = { index: self.atomTable.length(), atomicType };
                self.atomTable.add(result);
                return result;
            }
        }
    }
```

The BDD for `diff(M1,M2)` I get there is different. So there is a bug in `diff()` in jBallerina.

**Comment by lochana-chathura** [comment]
The root cause was found to be a bug in `mappingInhabited()` when porting. Both nBallerina and jBallerina produce the same BDDs. The previous BDD explained here turned out to be a mistake.

**Comment by lochana-chathura** [comment]
Fixed in PR #42791
